src/Conversion/ONNXToKrnl/Math/MatMul.cpp

```diff
@@ -315,10 +398,22 @@ struct ONNXMatMulOpLowering : public ConversionPattern {
     Value A(operandAdaptor.A()), B(operandAdaptor.B());
     auto aRank = A.getType().cast<MemRefType>().getShape().size();
     auto bRank = B.getType().cast<MemRefType>().getShape().size();
+    auto cRank = alloc.getType().cast<MemRefType>().getShape().size();
     if (enableTiling && aRank == 2 && bRank == 2) {
       // Optimized Matmul only when 2D and allowed to tile and unroll.
+      assert(cRank == 2 && "expected IxK * KxJ = IxJ 2D result");
       replace2x2Matmul2d(matMulOp, operandAdaptor, elementType, shapeHelper,
           alloc, zero, rewriter, loc);
+    } else if (enableTiling && aRank == 2 && bRank > 2) {
+      // Broadcasting B.
+      assert(cRank == bRank && "expected IxK * *xKxJ = *xIxJ result");
+      replace2x2Matmul2dBroadcasting(matMulOp, operandAdaptor, elementType,
+          shapeHelper, /*broadcasting B*/ true, alloc, zero, rewriter, loc);
+    } else if (enableTiling && aRank > 2 && bRank == 2) {
+      // Broadcasting A.
+      assert(cRank == aRank && "expected IxK * *xKxJ = *xIxJ result");
+      replace2x2Matmul2dBroadcasting(matMulOp, operandAdaptor, elementType,
+          shapeHelper, /*broadcasting B*/ false, alloc, zero, rewriter, loc);
```

Comment:
I think you can cover the case where `aRank == bRank && aRank > 2` with a small extension. It is not broadcasting but a batched matmul, e.g. `(B1xB2xIxK) * (B1xB2xKxJ)` and it does `(IxK) * (K*J)` for each batch index. Do you plan to do it in this patch also?

Reply:
When the B1...Bn are the same values? 

Because I think broadcasting also supports the case where a 1 is matched with a Bx and then there is a broadcasting there. The case you are suggesting is indeed easier, if they are equally matched. Is that what you had in mind?

Reply:
If it is, let me do in a different PR, because I also need to upgrade the ModelLib to generate and test for this new case. It does seems worthwhile.

Reply:
>  The case you are suggesting is indeed easier, if they are equally matched. Is that what you had in mind?

Yes, I am thinking about the "equally-matched" case at this moment.

Reply:
and no runtime, as there is no way to know if a `-1` will materialize as a `1` or matching '10' for example